src/udev/udev-builtin-dissect_image.c

```diff
@@ -0,0 +1,381 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+
+#include "blockdev-util.h"
+#include "device-util.h"
+#include "dissect-image.h"
+#include "fd-util.h"
+#include "hexdecoct.h"
+#include "image-policy.h"
+#include "initrd-util.h"
+#include "loop-util.h"
+#include "proc-cmdline.h"
+#include "udev-builtin.h"
+
+static ImagePolicy *arg_image_policy = NULL;
+static VeritySettings arg_verity_settings = VERITY_SETTINGS_DEFAULT;
+
+static int acquire_image_policy(ImagePolicy **ret) {
+        int r;
+
+        assert(ret);
+
+        _cleanup_free_ char *value = NULL;
+        r = proc_cmdline_get_key("systemd.image_policy", /* flags= */ 0, &value);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read systemd.image_policy= kernel command line switch: %m");
+        if (r == 0) {
+                *ret = NULL;
+                return 0;
+        }
+
+        r = image_policy_from_string(value, ret);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse image policy '%s': %m", value);
+
+        if (DEBUG_LOGGING) {
+                _cleanup_free_ char *s = NULL;
+
+                (void) image_policy_to_string(*ret, /* simplify= */ true, &s);
+                log_debug("Loaded image policy: %s", strna(s));
+        }
+
+        return 1;
+}
+
+static int acquire_verity_settings(VeritySettings *ret) {
+        _cleanup_(verity_settings_done) VeritySettings verity = VERITY_SETTINGS_DEFAULT;
+        int r;
+
+        assert(ret);
+
+        _cleanup_free_ char *h = NULL;
+        r = proc_cmdline_get_key("roothash", /* flags= */ 0, &h);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read roothash= kernel command line switch: %m");
+        if (r > 0)
+                verity.designator = PARTITION_ROOT;
+
+        _cleanup_free_ char *uh = NULL;
+        r = proc_cmdline_get_key("usrhash", /* flags= */ 0, &uh);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read usrhash= kernel command line switch: %m");
+        if (r > 0) {
+                if (h)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Both roothash= and usrhash= specified, refusing.");
+
+                h = TAKE_PTR(uh);
+                verity.designator = PARTITION_USR;
+        }
+
+        if (h) {
+                r = unhexmem(h, &verity.root_hash, &verity.root_hash_size);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse root hash from kernel command line switch: %m");
+        }
+
+        *ret = verity;
+        verity = VERITY_SETTINGS_DEFAULT;
```

Comment:
`TAKE_STRUCT()` ?

Reply:
doesn't really, work? TAKE_STRUCT() wants {} initialization, which we don#t want here.

Reply:
`ret = TAKE_GENERIC(verity, VeritySettings, VERITY_SETTINGS_DEFAULT);`

Maybe better to define a macro about that, as it is already used in core/namespace.c .